### PR TITLE
Monitor stack

### PR DESCRIPTION
**Pull request by SweBarre**
During the weekend I was playing around with prometheus and grafana.

I've created two files, `deploy/20.nginx-ingress.sh` and `deploy/21.monitor.sh`.
The reason for splitting them in two is that the `nginx-ingress` might be used for other deployments as well.

Here's some screenshots after deployments.

Screenshot from 2019-07-15 10-56-13
Screenshot from 2019-07-15 10-57-17

### CODE REVIEW

**Comment by ne0777** [comment]
Hey Jonas, I also put together some documentation on how to add the monitoring pieces to the deployment.  A couple of issues I see here.  First the PSP you have is basically working around the missing AppArmor components in the deployment.  We will be fixing(adding) the AppArmor components to the box so we don't have to disable it.  The new box will be ready after beta4 drops.

Next the dashboard json that you have embedded would be better as a separate file.  This will keep things tidier in the code and we could add them optionally.

Lastly the nginx.conf contains references to a bunch of workers that might not exist on every deployment.  Some may not have the resources to deploy that many workers.  And while I am not sure if errors would be logged for non-existent nodes (from a round-robin via nginx), it might confuse those who don't have these nodes.

Let's plan on collaborating around this a bit.

@sigsteve anything else you want to add?

**Comment by SweBarre** [comment]
> Hey Jonas, I also put together some documentation on how to add the monitoring pieces to the deployment. A couple of issues I see here. First the PSP you have is basically working around the missing AppArmor components in the deployment. We will be fixing(adding) the AppArmor components to the box so we don't have to disable it. The new box will be ready after beta4 drops.

Perfect, then I can remove this remark https://github.com/sigsteve/vagrant-caasp/pull/20/files#diff-83f925c4c6480fb6abc586e5ececdf15R4 :)

> 
> Next the dashboard json that you have embedded would be better as a separate file. This will keep things tidier in the code and we could add them optionally.

I'll fix, makes sense. 

> 
> Lastly the nginx.conf contains references to a bunch of workers that might not exist on every deployment. Some may not have the resources to deploy that many workers. And while I am not sure if errors would be logged for non-existent nodes (from a round-robin via nginx), it might confuse those who don't have these nodes.
> 

I agree, not sure the best way to handle this though.
I just followed the same configuration as for the master upstream, all potential masters were defined although you might just use one.
Suggestions would be appreciated.

File: cluster.sh [hunk 1]
```diff
@@ -133,6 +134,34 @@ function get_dashboard_credentials {
 
 }
 
+function get_monitoring_credentials {
+    local CAASP_DOMAIN="$(sed -n 's/^\s*domain\s*= "\(.*\)".*$/\1/p' Vagrantfile)"
+    cat << EOF
+You need to add the following to your /etc/hosts file:
+
+#vagrant-caasp4
+192.168.121.111     grafana.${CAASP_DOMAIN} prometheus.${CAASP_DOMAIN} prometheus-alert.${CAASP_DOMAIN}
```
Comment (sigsteve) [inline cluster.sh]: Can we add this to vagrant-libvirt.xml and the add hosts script?

File: boxfiles/setup_lb.sh [hunk 2]
```diff
@@ -1,3 +1,66 @@
...
+done
+printf "    }\n" >> /etc/nginx/nginx.conf
+
+printf "    upstream k8s-workers_https {\n" >> /etc/nginx/nginx.conf
+for NUM in $(seq 1 $NWORKERS); do
+    printf "        server caasp4-worker-1:443 weight=1 max_fails=1;\n" >> /etc/nginx/nginx.conf
+done
+cat >> /etc/nginx/nginx.conf << EOF
+    }
+
```
Comment (sigsteve) [inline boxfiles/setup_lb.sh]: Does it hurt to have multiple workers/masters defined, if they don't actually exist?
I left them all there initially to allow for easier expansion of the cluster (ie: add a new node) without needing to update the LB.

**Pull request merged**